routing/http/server/filters.go

```diff
@@ -0,0 +1,201 @@
+
+			if record != nil {
+				filtered = append(filtered, record)
+			}
+
+		} else {
+			// Will we ever encounter the SchemaBitswap type? Evidence seems to suggest that no longer
+			logger.Errorw("encountered unknown provider schema", "Schema", record.GetSchema(), "Type", reflect.TypeOf(record).String())
+		}
+	}
+	return filtered
+}
+
+// Applies the filters. Returns nil if the provider does not pass the protocols filter
+// The address filter is more complicated because it potentially modifies the Addrs slice.
+func applyFilters(provider *types.PeerRecord, filterAddrs, filterProtocols []string) *types.PeerRecord {
+	if len(filterAddrs) == 0 && len(filterProtocols) == 0 {
+		return provider
+	}
+
+	if !applyProtocolFilter(provider.Protocols, filterProtocols) {
+		// If the provider doesn't match any of the passed protocols, the provider is omitted from the response.
+		return nil
+	}
+
+	// return untouched if there's no filter or filterAddrsQuery contains "unknown" and provider has no addrs
+	if len(filterAddrs) == 0 || (len(provider.Addrs) == 0 && slices.Contains(filterAddrs, "unknown")) {
+		return provider
+	}
+
+	filteredAddrs := applyAddrFilter(provider.Addrs, filterAddrs)
+
+	// If filtering resulted in no addrs, omit the provider
+	if len(filteredAddrs) == 0 {
+		return nil
+	}
+
+	provider.Addrs = filteredAddrs
+	return provider
+}
+
+// If there are only negative filters, no addresses will be included in the result. The function will return an empty list.
+// For an address to be included, it must pass all negative filters AND match at least one positive filter.
+func applyAddrFilter(addrs []types.Multiaddr, filterAddrsQuery []string) []types.Multiaddr {
+	if len(filterAddrsQuery) == 0 {
+		return addrs
+	}
+
+	filteredAddrs := make([]types.Multiaddr, 0, len(addrs))
+
+	for _, addr := range addrs {
+		protocols := addr.Protocols()
+		includeAddr := true
+
+		// First, check all negative filters
+		for _, filter := range filterAddrsQuery {
+			if strings.HasPrefix(filter, "!") {
+				protocolStringFromFilter := strings.TrimPrefix(filter, "!")
+				protocolFromFilter := multiaddr.ProtocolWithName(protocolStringFromFilter)
+				if containsProtocol(protocols, protocolFromFilter) {
+					includeAddr = false
+					break
+				}
+			}
+		}
+
+		// If the address passed all negative filters, check positive filters
+		if includeAddr {
+			for _, filter := range filterAddrsQuery {
+				if !strings.HasPrefix(filter, "!") {
+					protocolFromFilter := multiaddr.ProtocolWithName(filter)
+					if containsProtocol(protocols, protocolFromFilter) {
+						filteredAddrs = append(filteredAddrs, addr)
+						break
+					}
+				}
+			}
+		}
+	}
+	return filteredAddrs
+}
+
+func containsProtocol(protos []multiaddr.Protocol, proto multiaddr.Protocol) bool {
+	for _, p := range protos {
+		if p.Code == proto.Code {
+			return true
+		}
+	}
+	return false
+}
+
+func applyProtocolFilter(peerProtocols []string, filterProtocols []string) bool {
```

Comment:
It is not immediately clear, without reading the oode, that returning true means the protocol is allowed. Perhaps rename this function to `protocolAllowed` or something similar.